### PR TITLE
Refactored ConvertableTo to allow for variadic declarations of pure virtual conversion functions

### PR DESCRIPTION
**Pull request by technetos**
### Changes proposed within this pull request:
- variadic declarations of pure virtual conversion functions
- classes that wish to have a value_type property should inherit from value_type_is directly
-  Stringafiable now inherits from value_type_is \<std::string\>.

### CODE REVIEW

**Comment by technetos** [comment]
The commit message got screwed up some how..anyways I put the whole commit message in the top comment.

File: C++/Logger/conversion/ConvertibleTo.h [hunk 1]
```diff
@@ -6,8 +6,11 @@ namespace Toolbox::Log::Conversion
 
 #include "../metaprogramming/value_type_is.h"
 
-template <typename ResultType>
-class ConvertibleTo : public value_type_is <ResultType>
```
Comment (DannyPeck) [inline C++/Logger/conversion/ConvertibleTo.h]: value_type_is should be declared on ConvertibleTo and not subclasses

File: C++/Logger/conversion/Stringifiable.h [hunk 2]
```diff
@@ -2,17 +2,20 @@
...
 
 #include <string>
 
 namespace Toolbox::Log::Conversion
 {
 
-class Stringifiable : public ConvertibleTo<std::string>
+class Stringifiable
+  : public ConvertibleTo<std::string>
+  , value_type_is <std::string>
```
Comment (DannyPeck) [inline C++/Logger/conversion/Stringifiable.h]: See comment above about value type

File: C++/Logger/conversion/Stringifiable.h [hunk 3]
```diff
@@ -2,17 +2,20 @@
...
 {
 public:
   virtual const std::string toString (void) const = 0;
-  operator std::string (void) const override final;
```
Comment (DannyPeck) [inline C++/Logger/conversion/Stringifiable.h]: The override should stay here, you can get rid of final but the override helps ensure that the signature of the method matches the inherited signature.

File: C++/Logger/tests/test_types/TestMessage.h [hunk 4]
```diff
@@ -7,15 +7,17 @@
...
 using Toolbox::Log::Conversion::ConvertibleTo;
 
-class TestMessage : public ConvertibleTo<std::string>
+class TestMessage : public ConvertibleTo<std::string, int>
 {
 public:
   explicit TestMessage (const std::string & message);
 
-  const operator std::string (void) const override;
+  operator std::string (void) const;
```
Comment (DannyPeck) [inline C++/Logger/tests/test_types/TestMessage.h]: See comment about override

File: C++/Logger/tests/conversion/ConvertibleTo.cpp [hunk 5]
```diff
@@ -12,4 +12,14 @@ BOOST_AUTO_TEST_CASE (convert_matches_input)
   const std::string convertedValue = testMessage;
 
   BOOST_TEST (convertedValue == message);
-}
\ No newline at end of file
+}
+
```
Comment (DannyPeck) [inline C++/Logger/tests/conversion/ConvertibleTo.cpp:16]: We should add tests for using the variatic template

**Review by DannyPeck — CHANGES_REQUESTED** [review]

File: C++/Logger/conversion/Stringifiable.h (hunk 3 above)
Comment (technetos) [reply C++/Logger/conversion/Stringifiable.h]: Its a pure virtual method, your not overriding anything.  The above **does not** compile.

File: C++/Logger/tests/conversion/ConvertibleTo.cpp (hunk 5 above)
Comment (technetos) [reply C++/Logger/tests/conversion/ConvertibleTo.cpp:16]: I did.

File: C++/Logger/conversion/ConvertibleTo.h (hunk 1 above)
Comment (technetos) [reply C++/Logger/conversion/ConvertibleTo.h]: `value_type` represents the parameterized type, in this case their are n >= 1 parameterized types.  We could make it the first one.

Comment (DannyPeck) [reply C++/Logger/conversion/ConvertibleTo.h]: On mobile so it is hard to review but doesn't your value_type_is have support for variatic types?

File: C++/Logger/conversion/Stringifiable.h (hunk 3 above)
Comment (DannyPeck) [reply C++/Logger/conversion/Stringifiable.h]: Stringifiable implements the conversion operator and calls toString so there is an actual implementation, therefore override is valid

File: C++/Logger/conversion/ConvertibleTo.h (hunk 1 above)
Comment (technetos) [reply C++/Logger/conversion/ConvertibleTo.h]: Oh right.

EDIT: No actually value_type supports nested types that can be variadic

File: C++/Logger/tests/conversion/ConvertibleTo.cpp (hunk 5 above)
Comment (DannyPeck) [reply C++/Logger/tests/conversion/ConvertibleTo.cpp:16]: My bad I missed that.

File: C++/Logger/conversion/Stringifiable.h (hunk 3 above)
Comment (technetos) [reply C++/Logger/conversion/Stringifiable.h]: Its an implementation of an interface.  There's no definition.  This is not method overriding.  On gcc this does not compile.  
```
$ make
g++ --std=c++1z -c conversion/Stringifiable.cpp
In file included from conversion/Stringifiable.cpp:1:0:
conversion/Stringifiable.h:18:31: error: expected ‘;’ at end of member declaration
   operator std::string (void) override const;
                               ^~~~~~~~
conversion/Stringifiable.h:18:40: error: declaration does not declare anything [-fpermissive]
   operator std::string (void) override const;
                                        ^~~~~
conversion/Stringifiable.h:18:3: error: ‘Toolbox::Log::Conversion::Stringifiable::operator std::string()’ marked ‘override’, but does not override
   operator std::string (void) override const;
   ^~~~~~~~
conversion/Stringifiable.cpp:6:53: error: expected initializer before ‘const’
 Stringifiable::operator std::string (void) override const
                                                     ^~~~~
makefile:36: recipe for target 'Stringifiable.o' failed
make: *** [Stringifiable.o] Error 1
```

File: C++/Logger/metaprogramming/type_is.h [hunk 6]
```diff
@@ -19,4 +19,7 @@ struct type_is <TT <T, Ts ...>>
   using type = T;
 };
 
+template <typename ... Ts>
+using type_is_t = typename type_is <Ts ...>::type;
```
Comment (xmclark) [inline C++/Logger/metaprogramming/type_is.h]: I like that you're taking a hint from the STL by using _t for type alias 😄

**Comment by xmclark** [comment]
Looks good.

File: C++/Logger/conversion/Stringifiable.h (hunk 3 above)
Comment (DannyPeck) [reply C++/Logger/conversion/Stringifiable.h]: Putting const after override is invalid syntax.  'const' qualifier may not appear after the virtual specifier 'override'.  Override helps you to identify subclasses that rely on that method and override it.  That way if you change the higher level interface, the classes that override it will not compile until you update them.

Comment (technetos) [reply C++/Logger/conversion/Stringifiable.h]: Why will the classes compile without the override specifier?? Also this really isnt method overriding.

Comment (DannyPeck) [reply C++/Logger/conversion/Stringifiable.h]: I think you maybe mistaking overriding with overloading.  Overriding is when a subclass reimplements a method defined in a base class thus providing new functionality.  The override specifier is optional, it adds a compile-time check that the function actually overrides a base method (so if you typed it wrong or the method signature changed it wouldn't compile).  This is helpful for identifying places in the code that need to be updated when you make an interface change and it is also helpful because it allows you to see which methods on a class override their base implementation.

Comment (DannyPeck) [reply C++/Logger/conversion/Stringifiable.h]: I see what you mean about there not being an implementation to begin with, but it will still have the above mentioned benefits and there really isn't a drawback.

Comment (technetos) [reply C++/Logger/conversion/Stringifiable.h]: ok.

**Review by DannyPeck — CHANGES_REQUESTED** [review]
Per Josh and I's conversion here is the list of concerns / things to test: We should move the first overload out of type_is because it will prevent us from converting to more complicated types (the first overload is essentially inner_type, e.g. inner_type<vector<int>>::type == int.  This will then result in ConvertibleTo<vector<int>> to resolve to int and not vector<int> which is not expected or desired behavior.

We should also test the variadic specialization for at a minimum the following scenarios: passing in the same type twice and passing in both a base and derived type from the same class hierarchy.

**Review by DannyPeck — APPROVED** [review]
